### PR TITLE
Build optimizations

### PR DESCRIPTION
**Pull request by hekaldama**

### CODE REVIEW

**Comment by tsandall** [comment]
hey @hekaldama, thanks for looking into this! We definitely want to work towards TLS support on the server-side.

Could you provide details on the motivation for all of these changes? Just update the PR description or add a comment. I'll review the rest of the PR in the mean time.

That being said, a couple things jump out at me:

- Disabling cgo is fine, but it doesn't give us a cross-platform executable--the executable will not have any dynamic library dependencies but the executable format is still platform specific (e.g., ELF on Linux, Mach-O on macOS, etc.) We still have to produce multiple binaries (hence the current Makefile structure).

- Committing the executable into the repo is not desirable.

- How much space is saved by stripping the debug symbols? If we're only talking about a few MBs, I think it's worth keeping them in.

**Comment by hekaldama** [comment]
Sure thing. I am new here wrt some of this stuff, so happy to collaborate on best practices. For some reason my commit message (https://github.com/open-policy-agent/opa/pull/125/commits/fcc8de3d35afe001d505d631c6b0a97ae78bb16d) didn't come through well in the PR.

In response to your bullets above:

* Ah! Good to know. I thought if its static, then it works everywhere. Shows my lack of understanding around cross platform compilation.

* I put the executable in there so hub.docker.com can autobuild your tags without having to have a full blown golang docker image. The workflow I am envisioning is build the binary(ies) locally, tag the release, push to github, hub.docker.com autobuilds tags and the Dockerfile just adds the binary to the image and publishes the image. I am unsure if this is the best way, but best way I could think of.

* Stripping symbols saves 3.95MB. Again, I am naive here, what do symbols help with?

Thanks.

File: Dockerfile [hunk 1]
```diff
@@ -2,12 +2,18 @@
 # Use of this source code is governed by an Apache2
 # license that can be found in the LICENSE file.
 
-FROM scratch
+FROM alpine:latest
 
 MAINTAINER Torin Sandall <torinsandall@gmail.com>
 
-ADD /opa_linux_GOARCH /opa
+ENV BIN /usr/bin/opa
```
Comment (tsandall) [inline Dockerfile]: [minor nitpick] Is /usr/bin the right place for this? I'm not super familiar with alpine.

File: Dockerfile [hunk 2]
```diff
@@ -2,12 +2,18 @@
...
 
-ENTRYPOINT ["/opa"]
+RUN apk add --update ca-certificates &&\
+    rm -rf /var/cache/apk/* &&\
+    update-ca-certificates
 
-CMD ["run"]
```
Comment (tsandall) [inline Dockerfile]: Is there a particular reason for this? The default command was specified so that `docker run -i --rm openpolicyagent/opa` just did the right thing.

File: Makefile [hunk 3]
```diff
@@ -31,7 +29,8 @@ BUILD_HOSTNAME := $(shell ./build/get-build-hostname.sh)
 LDFLAGS := "-X github.com/open-policy-agent/opa/version.Version=$(VERSION) \
 	-X github.com/open-policy-agent/opa/version.Vcs=$(BUILD_COMMIT) \
 	-X github.com/open-policy-agent/opa/version.Timestamp=$(BUILD_TIMESTAMP) \
-	-X github.com/open-policy-agent/opa/version.Hostname=$(BUILD_HOSTNAME)"
+	-X github.com/open-policy-agent/opa/version.Hostname=$(BUILD_HOSTNAME) \
+	-s"
```
Comment (tsandall) [inline Makefile]: Symbols are typically used for stack traces. There are also DWARF structures for debugging. It looks like you can disable that as well ("-w"). That being said, at this stage in the project, it's probably good to keep both on by default. If size really is a concern, an optional input parameter could be used.

File: Makefile [hunk 4]
```diff
@@ -47,12 +46,10 @@ generate:
 	$(GO) generate
 
 build: generate
-	$(GO) build -o $(BIN) -ldflags $(LDFLAGS)
+	CGO_ENABLED=0 $(GO) build -o $(BIN) -ldflags $(LDFLAGS)
```
Comment (tsandall) [inline Makefile]: Disabling CGO is probably fine for now. If we need it in the future, we can turn it back on.

Comment (tsandall) [inline Makefile]: Do we need CGO_ENABLED=0 on the install and test targets as well?

**Comment by tsandall** [comment]
I left a few comments on the changes. It would be helpful to break these changes into smaller PRs (that each have a specific purpose).

- Disabling CGO is fine and could be a separate PR.

- Switching to alpine is fine if it's the best way of supporting TLS on the server-side (though I'm not sure this is the case.) This should be a separate PR.

As far as automated releases go, have you looked at https://github.com/open-policy-agent/opa/blob/master/docs/RELEASE.md? The current process requires manual steps to mark the release and then publish the artifacts (e.g., binaries, docker images, site updates, etc.) If there's a way of automating the publishing step without requiring the binaries be checked into the repo, that would be great.

We currently use Travis CI to build all PRs before they're merged. I wonder if we could also use it to automatically build & publish tags?

File: Dockerfile (hunk 1 above)
Comment (hekaldama) [inline Dockerfile]: I don't know, but I also don't think it matters much because in a docker world, you can control a lot of system things and don't have to worry much about other tasks (cron, package updates, etc.) messing you up. I am sure people are more passionate about where to put it and really don't mind where it goes. I just liked it in the `$PATH` :).

File: Dockerfile (hunk 2 above)
Comment (hekaldama) [inline Dockerfile]: As a non-dev of the project I liked it to tell me usage instead of just running. Similar to if I do `curl`, `docker`, `less`, etc. I realize `bash` runs it for you and doesn't give you usage, so again, I could go either way.

File: Makefile (hunk 4 above)
Comment (hekaldama) [inline Makefile]: Good question. I would say probably, but can research more.

**Comment by hekaldama** [comment]
Thanks for the feedback. I will take a look at breaking this apart into other PRs, and automated build setups next week. I will close and comment.

File: Dockerfile [hunk 5]
```diff
@@ -2,12 +2,18 @@
 # Use of this source code is governed by an Apache2
 # license that can be found in the LICENSE file.
 
-FROM scratch
+FROM alpine:latest
```
Comment (tsandall) [inline Dockerfile]: Ah, OK, the reason it's failing is because /tmp doesn't exist in the scratch image. It should be fairly easy to work around this (either create it at build time, or modify OPA to create the directory on startup if it doesn't exist). That being said, logging to stderr seems like the way to go, e.g., no need to do rotation, worry about disk overflowing, can easily hook up to syslog, etc.

Right now there wouldn't be much gained by having a shell inside the container--I know what you're describing, and it's definitely valuable for heavy-weight images, but in this case, I haven't found a need.

Based on this, I'm not sure we gain much. Maybe it's easier to support TLS?

File: Dockerfile (hunk 2 above)
Comment (tsandall) [inline Dockerfile]: Ah, that makes sense. I always think of the default behaviour as the REPL but printing the command help is probably better. Let's make that change in a separate PR.
